tests/compiler/std/array.optimized.wat

```diff
@@ -3328,15 +3334,239 @@
+   i32.ge_s
+  else
+   i32.const 1
+  end
+  if
+   i32.const -1
+   return
+  end
+  local.get $0
+  i32.load offset=4
+  local.set $0
+  loop $continue|0
+   local.get $1
+   local.get $2
+   i32.lt_s
+   if
+    local.get $1
+    i32.const 2
+    i32.shl
+    local.get $0
+    i32.add
+    f32.load
+    f32.const nan:0x400000
+    f32.eq
+    if
+     local.get $1
+     return
+    else
+     local.get $1
+     i32.const 1
+     i32.add
+     local.set $1
+     br $continue|0
+    end
+    unreachable
+   end
+  end
+  i32.const -1
+ )
+ (func $~lib/array/Array<f64>#indexOf (; 58 ;) (type $FUNCSIG$ii) (param $0 i32) (result i32)
+  (local $1 i32)
+  (local $2 i32)
+  local.get $0
+  i32.load offset=12
+  local.tee $2
+  if (result i32)
+   i32.const 0
+   local.get $2
+   i32.ge_s
+  else
+   i32.const 1
+  end
+  if
+   i32.const -1
+   return
+  end
+  local.get $0
+  i32.load offset=4
+  local.set $0
+  loop $continue|0
+   local.get $1
+   local.get $2
+   i32.lt_s
+   if
+    local.get $1
+    i32.const 3
+    i32.shl
+    local.get $0
+    i32.add
+    f64.load
+    f64.const nan:0x8000000000000
+    f64.eq
+    if
+     local.get $1
+     return
+    else
+     local.get $1
+     i32.const 1
+     i32.add
+     local.set $1
+     br $continue|0
+    end
+    unreachable
+   end
+  end
+  i32.const -1
+ )
+ (func $~lib/array/Array<i32>#includes (; 59 ;) (type $FUNCSIG$iiii) (param $0 i32) (param $1 i32) (param $2 i32) (result i32)
   local.get $0
   local.get $1
   local.get $2
   call $~lib/array/Array<i32>#indexOf
   i32.const 0
   i32.ge_s
  )
- (func $~lib/array/Array<i32>#splice (; 58 ;) (type $FUNCSIG$iiii) (param $0 i32) (param $1 i32) (param $2 i32) (result i32)
+ (func $~lib/number/isNaN<f32> (; 60 ;) (type $FUNCSIG$if) (param $0 f32) (result i32)
+  local.get $0
+  local.get $0
+  f32.ne
+ )
+ (func $~lib/array/Array<f32>#includes (; 61 ;) (type $FUNCSIG$ii) (param $0 i32) (result i32)
+  (local $1 i32)
+  (local $2 i32)
+  (local $3 f32)
+  local.get $0
+  i32.load offset=12
+  local.tee $2
+  if (result i32)
+   i32.const 0
+   local.get $2
+   i32.ge_s
+  else
+   i32.const 1
+  end
+  if
+   i32.const 0
+   return
+  end
+  local.get $0
+  i32.load offset=4
+  local.set $0
+  loop $continue|0
+   local.get $1
+   local.get $2
+   i32.lt_s
+   if
+    local.get $1
+    i32.const 2
+    i32.shl
+    local.get $0
+    i32.add
+    f32.load
+    local.tee $3
+    f32.const nan:0x400000
+    f32.eq
+    if (result i32)
+     i32.const 1
+    else
+     local.get $3
+     call $~lib/number/isNaN<f32>
+     f32.const nan:0x400000
+     call $~lib/number/isNaN<f32>
+     i32.and
+    end
+    if
+     i32.const 1
+     return
+    else
+     local.get $1
+     i32.const 1
+     i32.add
+     local.set $1
+     br $continue|0
+    end
+    unreachable
+   end
+  end
+  i32.const 0
+ )
+ (func $~lib/number/isNaN<f64> (; 62 ;) (type $FUNCSIG$id) (param $0 f64) (result i32)
+  local.get $0
+  local.get $0
+  f64.ne
+ )
```

Comment:
@dcodeIO We definitely need tune inlining limits more better than binaryen's defaults.

Reply:
Hmm, that's interesting. From looking over the inlining limits, this one seems as if it is smaller than `flexibleInlineMaxSize = 20` and should be considered `lightweight` (maybe it isn't?), so it should inline if `optimizeLevel >= 3` and `shrinkLevel == 0`.

Reply:
Yes, in `-O3` mot `-O3z` it will be inlined but I guess even with `shrinkLevel == 1` this should be possible. Only for `shrinkLevel == 2 and 3` this unnecessary I guess

Reply:
btw fan fact:
`-O3` example with four `isNaN`: https://webassembly.studio/?f=tsbtqbkt2ks
has size: `123 bytes` and with inlines

Same example but with `-O3z`: https://webassembly.studio/?f=4jfqpshboq9
`142 bytes`

Reply:
I guess sometimes, when code is inlined, even if it doesn't meet the inlining criteria introduced to avoid growing, it can be optimized much better further down the road than if it remained its own small function. The challenge here seems to be to detect that condition, which would have to estimate how much inlining a function that isn't considered worth inlining at first actually benefits from other passes after inlining. Quite a hard problem, that doesn't seem to be solvable by just tweaking the inlining options because this would then also inline stuff that doesn't actually benefit from other passes. As such, I think Binaryen is doing what it reasonably can do without either recompiling multiple times to observe the difference or go completely crazy trying to build an estimator.

Reply:
Maybe one assumption we could apply in `Os` is to ignore pushes to the stack involving function arguments, like the `local.get`s in this specific example, expecting that these will *usually* blend in fine into the call site? But I'm guessing.